### ISSUE TITLE
(Likely) false positive for unnecessary defer

### ISSUE THREAD

**Issue by ferhatelmas** [go-critic/go-critic#941]
```go
package main

import (
	"fmt"
	"sync"
)

type T struct {
	l    *sync.Mutex
	data []int
}

func (t *T) Index(i int) int {
	t.l.Lock()
	defer t.l.Unlock()
	return t.data[i]
}

func main() {
	t := T{l: &sync.Mutex{}, data: []int{1, 2, 3}}
	fmt.Println(t.Index(2))
}
```

Implementation of the index method is idiomatic and having this warning in this case isn't optimum.

```sh
 gocritic check -enable='unnecessaryDefer' main.go
./main.go:15:2: unnecessaryDefer: defer t.l.Unlock() is placed just before return
```

**Comment by cristaloleg** [comment]
Hi @ferhatelmas can you share your go-critic's version, please?

**Comment by ferhatelmas** [comment]
@cristaloleg I realized it with the latest golangci-lint (`v1.28.1`) and also tested now with gocritic `v0.5.0`.

**Comment by quasilyte** [comment]
https://github.com/go-critic/go-critic/blob/master/checkers/testdata/unnecessaryDefer/negative_tests.go

We lack a test for this case.

I would propose this fix:
* Check that all returned "arguments" are trivial (consts, literals, stuff like that) and don't need synchronization.
* If the return contains anything we don't understand, skip this case and don't report unnecessary defer warning
